segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -0,0 +1,510 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.pravega.segmentstore.storage.metadata;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import lombok.Builder;
+import lombok.Data;
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+
+/**
+ * Implements base metadata store that provides core functionality of metadata store by encapsulating underlying key value store.
+ * Derived classes override {@link BaseMetadataStore#read(String)} and {@link BaseMetadataStore#writeAll(Collection)} to write to underlying storage.
+ * The minimum requirement for underlying key-value store is to provide optimistic concurrency ( Eg. using versions numbers or etags.)
+ *
+ * Within a segment store instance there should be only one instance that exclusively writes to the underlying key value store.
+ * For distributed systems the single writer pattern must be enforced through external means. (Eg. for table segment based implementation
+ * tier-1 fencing is used to establish ownership.
+ *
+ * This implementation provides following features that simplify metadata management.
+ *
+ * 1. It provides simple transactions API with snapshot isolation that is independent of underlying key-value store API .
+ * Changes made to metadata inside a transaction are not visible until a transaction is committed.
+ * In addition they are atomic - either all changes in the transaction are committed or none at all.
+ * Snapshot isolation means that transaction is fails if any of the metadata records read and during the transactions are changed after they were read.
+ *
+ * 2. It buffers frequently or recently updated metadata keys to optimize read/write to underlying.
+ * Especially it provides "lazy committing" of changes where there is application specific way to recover from failures.(Eg when only length of chunk is changed.)
+ * (Note that otherwise for each commit the data is written to underlying storage.)
+ *
+ */
+@Slf4j
+abstract public class BaseMetadataStore implements ChunkMetadataStore {
+    /**
+     * Maximum number of metadata entries to keep in recent transaction buffer.
+     */
+    private static final int MAX_ENTRIES_IN_TXN_BUFFER = 5000; //TODO : load from the config
+
+    /**
+     * List of pinned keys. These keys should not be read or written.
+     */
+    protected final ConcurrentSkipListSet<String> pinnedKeys = new ConcurrentSkipListSet<>();
```

Comment:
Do not expose your inner data structures, even as protected. Write APIs to access and/or mutate them.

Reply:
Same with the next few fields.